### ISSUE TITLE
False negative on function call with conditional type in loop

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#4254]
**Describe the bug**
Calling a function with a local variable that was reassigned using some condition in a loop, no errors are emitted although the argument type does not match.
Originally noticed this when a specific `# type: ignore` was seemingly no longer needed using the latest version.

**To Reproduce**
See code below.
The loop, initial assignment, and condition all seem to be required for this to surface.

It took me a while to compress the original code down to an MCVE, this is about as far as I could get it. It doesn't really make sense, but I hope it helps illustrate the issue.

**Expected behavior**
`Argument of type "int" cannot be assigned to parameter "arg" of type "str" in function "g"`

**Screenshots or Code**
```python
def f(arg: int):
    for _ in range(1):
        loc = arg
        loc = loc if loc else loc

        reveal_type(loc)  # > Type of "loc" is "int"
        g(loc)  # no error

def g(arg: str):
    print(type(arg))

f(0)  # prints `<class 'int'>`
```

*Edit: removed `str()` cast, which isn't needed to reproduce this*

**VS Code extension or command-line**
cli, 1.1.282

**Additional context**
bisecting shows that this was introduced by cec5d7d5cb7e7b9c97ec5a5dcda7cd62c0995254/#4110, which was the fix for a previous issue I filed - sorry for causing headaches :sweat_smile:

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.283, which I just published. It will also be included in a future version of pylance.
